### ISSUE TITLE
mongoose has assignability error after #42248

### ISSUE THREAD

**Issue by sandersn** [microsoft/TypeScript#42337]
To repro: `cd DefinitelyTyped/types/mongoose && tsc`; observe that there are 74 errors, 2 more than the expected 72.

```ts
const locationSchema = new mongoose.Schema({
  name: { type: String, required: true },
  address: String,
  rating: { type: Number, "default": 0, min: 0, max: 5 },
  facilities: [String],
  coords: { type: [Number], index: "2dsphere" },
  openingTimes: [mongoose.Schema.Types.Mixed],
  reviews: [mongoose.SchemaTypes.Mixed],
  notes: [noteSchema]
});
var LocModel = mongoose.model<Location>("Location", locationSchema);
LocModel.create({ address: "foo", coords: [1, 2], facilities: ["foo", "bar"], name: "bar", openingTimes: ["foo"], rating: 10, reviews: ["foo"], notes: [] });
```

Actual:

At mongoose/test/model.ts:613, errors on openingTimes and reviews:

```
ERROR: 613:107  expect  TypeScript@4.2 compile error: 
No overload matches this call.
  The last overload gave the following error.
    Type 'string' is not assignable to type 'never'.
      Type 'string' is not assignable to type 'never'.
```

Expected: No error; the input type was `[mongoose.SchemeTypes.Mixed]`, which should result in `string[]` not `never[]`

Likely related to #42336, although in this case we can't ship with this error in mongoose, which may involve fixing mongoose in some way.

**Comment by xandris** [comment]
Minimal repro:

```typescript
type T = any extends number ? 1 : 0;
let x: T;
x = 1;
x = 0;
```

This used to typecheck, but now `T` is `1`. The commit that changed this behavior is:

```
c3dd845923c34cf12cc00cc0787556c177607507 is the first bad commit
commit c3dd845923c34cf12cc00cc0787556c177607507
Author: Wesley Wigham <wewigham@microsoft.com>
Date:   Tue Jan 12 12:59:52 2021 -0800

    Better detect when typical nondistributive conditionals need to be defered by unwrapping their check
 and extends types (#42248)
```

**Comment by xandris** [comment]
This relates to mongoose because `mongoose.DeepMapAsObject<{foo: any[]}>` used to result in `{foo: any[]}` but now results in `{foo: never[]}`. Here's the definition:

```typescript
type DeepMapAsObject<T> = T extends object | undefined
? {
    [K in keyof T]: T[K] extends Map<infer KM, infer KV> | undefined
    // if it's a map, transform it into Map | Record
    // only string keys allowed
    ? KM extends string ? Map<KM, DeepMapAsObject<KV>> | Record<KM, DeepMapAsObject<KV>> | [KM, DeepMapAsObject<KV>][] : never
    // otherwise if it's an object or undefined (for optional props), recursively go again
    : T[K] extends object | undefined
      ? DeepMapAsObject<T[K]>
      : T[K]
    }
: T;
```

Substituting `any[]` for `T`:

```
type DeepMapAsObject<any[]> = { /* since any[] extends object */
    // keyof any[] is number, any[number] is any
    [number]: any extends Map<unknown, unknown> | undefined // these get inferred as unknown
    // if it's a map, transform it into Map | Record
    // only string keys allowed
    ? never // KM is unknown which does not extend string, result is never
    // otherwise if it's an object or undefined (for optional props), recursively go again
    : any extends object | undefined
      ? DeepMapAsObject<any> // not too clear what happens here...
      : any
    };
```

So you have that conditional type testing whether `any extends`, and one branch is `never` and the other is `any`. Before the commit mentioned, the result was `{ [K in number]: any | never}` which is just `{ [K in number]: any}`. After, it just takes the "true" branch and the type becomes `{ [K in number]: never }`.
